source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -152,74 +173,80 @@ static int8_t ns_block_validate(int *block_start, int direction)
+                int block_data_size = -(*ptr);
+
+                if (block_data_size >= data_size) {
+                    /*found block*/
+                    if (block_data_size >= (data_size + 2 + HOLE_T_SIZE)) {
+                        int hole_size = block_data_size - data_size - 2;
+
+                        //There is enough room for a new hole so create it first
+                        if ( direction > 0 ) {
+                            hole_ptr = ptr + data_size + 2;
+                        } else {
+                            hole_ptr = ptr;
+                        }
+                        if( direction > 0 ){
+                            // Would like to just replace this cur_hole with new descriptor, but
+                            // they could overlap, so ns_list_replace might fail
+                            //ns_list_replace(&holes_list, cur_hole, hole_from_block_start(hole_ptr));
+                            hole_t *before = ns_list_get_previous(&holes_list, cur_hole);
+                            ns_list_remove(&holes_list, cur_hole);
+                            if (before) {
+                                ns_list_add_after(&holes_list, before, hole_from_block_start(hole_ptr));
                             } else {
-                                hole_ptr -= (hole_size + 1);
+                                ns_list_add_to_start(&holes_list, hole_from_block_start(hole_ptr));
                             }
-                            *hole_ptr = -(hole_size);
-                        }else{
-                            data_size = block_data_size;
                         }
 
-                        *ptr = data_size;
-                        if (direction > 0) {
-                            ptr++;
-                            retval = ptr;
-                            ptr += data_size;
-                        } else {
-                            ptr -= data_size;
-                            retval = ptr;
-                            ptr--;
+                        *hole_ptr = -(hole_size);
+                        hole_ptr += (hole_size + 1);
+                        *hole_ptr = -(hole_size);
+                        hole_ptr -= hole_size;
+                        if( direction < 0 ){
+                            ptr += hole_size + 2;
+                            hole_ptr = NULL; //no need to update
                         }
-                        *ptr = data_size;
-                        break;
+                    } else {
+                        data_size = block_data_size;
+                        ns_list_remove(&holes_list, hole_from_block_start(ptr));
                     }
-                }
-
-                moved += (block_data_size + 2);
-                if (direction > 0) {
-                    ptr += block_data_size + 2;
-                } else {
-                    ptr -= (block_data_size + 2);
+                    *ptr = data_size;
```

Comment:
Again,

          ptr[0] = +data_size;
          ptr[1 + data_size] = +data_size;
          retval = ptr + 1;

Not a fan of this "tape-seeking" behaviour on a single block.